service/storage_proxy.cc

```diff
@@ -1177,10 +1177,44 @@ future<bool> paxos_response_handler::accept_proposal(const paxos::proposal& prop
 future<> paxos_response_handler::learn_decision(paxos::proposal decision, bool allow_hints) {
     tracing::trace(tr_state, "learn_decision: committing {} with cl={}", decision, _cl_for_learn);
     paxos::paxos_state::logger.trace("CAS[{}] learn_decision: committing {} with cl={}", _id, decision, _cl_for_learn);
-    std::array<std::tuple<paxos::proposal, schema_ptr, dht::token>, 1> m{std::make_tuple(std::move(decision), _schema, _key.token())};
     // FIXME: allow_hints is ignored. Consider if we should follow it and remove if not.
     // Right now we do not store hints for when committing decisions.
-    return _proxy->mutate_internal(std::move(m), _cl_for_learn, false, tr_state, _permit, _timeout);
+    future<> f_cdc = make_ready_future<>();
+    if (_schema->cdc_options().enabled()) {
+        auto update_mut = decision.update.unfreeze(_schema);
+        const auto base_tbl_id = update_mut.column_family_id();
+        std::vector<mutation> update_mut_vec;
+        update_mut_vec.push_back(std::move(update_mut));
+
+        if (_proxy->get_cdc_service() && _proxy->get_cdc_service()->needs_cdc_augmentation(update_mut_vec)) {
+            // CDC mutations are sent down another route
+            f_cdc = _proxy->get_cdc_service()->augment_mutation_call(_timeout, std::move(update_mut_vec))
+                    .then([this, base_tbl_id] (std::tuple<std::vector<mutation>, cdc::result_callback>&& t) {
+                auto mutations = std::move(std::get<0>(t));
+                auto end_func = std::move(std::get<1>(t));
+
+                // Pick only the CDC ("augmenting") mutations
+                mutations.erase(std::remove_if(mutations.begin(), mutations.end(), [base_tbl_id = std::move(base_tbl_id)] (const mutation& v) {
+                    return v.schema()->id() == base_tbl_id;
+                }), mutations.end());
+                if (mutations.empty()) {
+                    return make_ready_future<>();
+                }
+
+                auto f = _proxy->mutate_internal(std::move(mutations), _cl_for_learn, false, tr_state, _permit, _timeout);
+                if (end_func) {
+                    f = f.then(std::move(end_func));
+                }
+                return f;
+            });
+        }
+    }
+
+    // Path for the "base" mutations
```

Comment:
Can this part of the code - the "base" (normal) write - be moved before the special CDC code. I think it will be clearer, and in any case, both operations happen in parallel. I think it would make the code clearer - you see the normal code path first, then the special cdc path. 
UPDATE: this suggestion is an alternative to the other idea, of not having separate with-cdc and without-cdc paths - both will generate a vector of mutations with a single mutate_internal() call at the end.

Reply:
I don't think it's possible to reorder the cases because execution of the lwt logic will destroy the mutation that we need to execute cdc logic. We could copy the mutation before executing lwt and use it later in cdc but then we would have:
1. cdc
2. lwt
3. cdc

which in my opinion is worse. What do you think @nyh ?

Reply:
The "1. cdc" part will just be an unfreeze() call, but you're right, that isn't very pretty either. I guess we can think about this later, and this version is acceptable. I'll commit.

Reply:
Thanks!